Packs/Akamai_SIEM/Integrations/Akamai_SIEM/Akamai_SIEM.py

```diff
@@ -551,6 +558,621 @@ def decode_url(headers: str) -> dict:
+        chunk = data[i:i + num_of_entries_per_chunk]
+        yield chunk
+
+
+async def xsiam_api_call_async_with_retries(
+    xsiam_url,
+    zipped_data,
+    headers,
+    num_of_attempts,
+    data_type=EVENTS,
+):  # pragma: no cover
+    """
+    Send the fetched events or assets into the XDR data-collector private api.
+    :type xsiam_url: ``str``
+    :param xsiam_url: The URL of XSIAM to send the api request.
+    :type zipped_data: ``bytes``
+    :param zipped_data: encoded events
+    :type headers: ``dict``
+    :param headers: headers for the request
+    :type num_of_attempts: ``int``
+    :param num_of_attempts: The num of attempts to do in case there is an api limit (429 error codes).
+    :type data_type: ``str``
+    :param data_type: events or assets
+    :return: Response object or DemistoException
+    :rtype: ``requests.Response`` or ``DemistoException``
+    """
+    # retry mechanism in case there is a rate limit (429) from xsiam.
+    status_code = None
+    attempt_num = 1
+    response = None
+    while status_code != 200 and attempt_num < num_of_attempts + 1:
+        demisto.debug('Sending {data_type} into xsiam, attempt number {attempt_num}'.format(
+            data_type=data_type, attempt_num=attempt_num))
+        # in the last try we should raise an exception if any error occurred, including 429
+        ok_codes = (200, 429) if attempt_num < num_of_attempts else None
+        async with aiohttp.ClientSession() as session:
+            async with session.post(urljoin(xsiam_url, '/logs/v1/xsiam'), data=zipped_data, headers=headers) as response:
+                try:
+                    response.raise_for_status()  # This raises an exception for non-2xx status codes
+                    status_code = response.status
+                except aiohttp.ClientResponseError as e:
+                    if ok_codes and e.status in ok_codes:
+                        continue
+                    else:
+                        header_msg = 'Error sending new {data_type} into XSIAM.\n'.format(data_type=data_type)
+                        api_call_info = (
+                            'Parameters used:\n'
+                            '\tURL: {xsiam_url}\n'
+                            '\tHeaders: {headers}\n\n'
+                            'Response status code: {status_code}\n'
+                            'Error received:\n\t{error}\n'
+                            'additional request info: \n\t{additional_info}'
+                        ).format(xsiam_url=xsiam_url, headers=json.dumps(e.headers, indent=8), status_code=e.status,
+                                 error=e.message, additional_info=e.request_info)
+
+                        demisto.error(header_msg + api_call_info)
+                        demisto.updateModuleHealth(header_msg + e.message, is_error=True)
+
+        demisto.debug('received status code: {status_code}'.format(status_code=status_code))
+        if status_code == 429:
```

Comment:
do we still have the logic that if we hit 429 we will redo the offset? We should still keep this in otherwise customers wont be able to pause the integration and have it naturally recover

Reply:
I think you're confusing with 416 in the request to Akamai.
This is 429 when sending events to xsiam.

Reply:
true. lets just make sure we still have that mechanism

Reply:
We do.